### DISCUSSION TITLE
Accept connections from only one client

### DISCUSSION THREAD

**Discussion by steteb** [cesanta/mongoose#1619] [Ask questions, get answers] [answered]
Hello,

I would like to know if it is possible to make Mongoose accept connections from only one client at a time ? So if there is already a client connected, Mongoose should not accept connections from other clients. For completeness I add that I am using Mongoose on a STM32H753 with FreeRTOS and LwIP.

Thank you

**Comment by cpq** [comment, marked as answer]
In the MG_EV_ACCEPT handler, traverse c->mgr->conns, and if you see a client already connected, mg_error().

Why? Is it a free RAM issue?

**Reply by steteb** [reply]
Hello, 

thanks.

How can I check that a client is already connected ? If I check is_accepted, it is true for the current accepted connection.

No, not only for the RAM, but also because it is a request for our application

**Reply by scaprile** [reply]
_**traverse c->mgr->conns**_, something like
```c
  } else if (ev == MG_EV_ACCEPT) {
    for (struct mg_connection *cc = c->mgr->conns; cc != NULL; cc = cc->next) {
      if ((cc != c) && (cc->is_accepted)) {
        mg_error(c,"many too many");
        break;
      }
    }
  }
```

**Reply by steteb** [reply]
Hello,

thanks.

Ok, yes, then I did that, but it doesn't seem to work. The client is probably opening multiple connections to Mongoose. Is there no way to identify the connected client (ip, port) ?

**Reply by scaprile** [reply]
If your application is a web server, yes, browsers do open several connections in parallel, you can see that sniffing your network with Wireshark. Each connection will have a different port number even if they come from the same browser window and tab. As for the IP, several browsers on the same machine (2 tabs, 2 browsers, a multi-user machine...) or on different machines behind NAT, will present you the same IP.
If I understand you correctly, you want the concept of _session_; that is what tokens and cookies do. Check the [device dashboard tutorial](https://mongoose.ws/tutorials/device-dashboard/) and see if that helps you in achieving your goal

**Reply by steteb** [reply]
Hello,

thank you very much. Very interesting the tutorial you suggested.

**Reply by cpq** [reply]
As @scaprile mentioned, browsers makes several requests:
1. They fetch index.html first
2. They fetch an icon, to display in a browser tab. If index.html has an explicit `<link rel="icon" href="..." type="image/x-icon" />` then it fetches it, otherwise it fetches /favicon.ico
3. Then for CSS/JS/image local references present in the index.html, browsers start a pool of 4-5 connections to fetch those resources as fast as possible

Overall, for a simple web page, browsers make a burst of connections which then are closed in a couple of minutes in case of inactivity. But the initial burst could be a problem - especially if TLS is used.

To alleviate , you might amalgamate your page into a single one with no references, by inlining all JS and CSS files. You can convert all your images into SVG and inline them too - or alternatively, use PNG and inline as base64 data. This way you can get away with a single connection instead of multiple.
